primitives/account/src/account/staking_contract/staker.rs

```diff
@@ -637,6 +770,106 @@ impl StakingContract {
+        let new_retired_balance = staker.retired_balance - value;
+        if !Self::is_min_stake_enforced(
+            staker.balance,
+            staker.inactive_balance,
+            new_retired_balance,
+        ) {
+            debug!(
+                active_balance=?staker.balance,
+                inactive_balance=?staker.inactive_balance,
+                retired_balance=?staker.retired_balance,
+                ?value,
+                "Tried to remove stake that would violate the minimum total stake"
+            );
+            return Err(AccountError::InvalidCoinValue);
+        }
+
+        Ok(())
+    }
+
+    /*  Helpers for checking release block heights and invariants */
+
+    /// Checks if the minimum stake invariants are respected.
+    /// This function expects as input the final balances of a given operation.
+    /// Invariants:
+    ///         (1) active + inactive balances should be == 0 or >= Minimum Stake
+    ///         (2) active + inactive + retired balances should be == 0 or >= Minimum Stake
+    fn is_min_stake_enforced(
+        active_balance: Coin,
+        inactive_balance: Coin,
+        retired_balance: Coin,
+    ) -> bool {
+        // Check invariant 1: that total non retired balance (active+inactive) does not violate the minimum stake.
+        let non_retired_balances = active_balance + inactive_balance;
+        if non_retired_balances > Coin::ZERO
+            && non_retired_balances < Coin::from_u64_unchecked(Policy::MINIMUM_STAKE)
+        {
+            return false;
+        }
+
+        // Check invariant 2: total stake does not violate minimum stake.
+        // If all funds are retired, then retired funds must be 0 or at least minimum stake.
+        if non_retired_balances.is_zero()
+            && retired_balance > Coin::ZERO
+            && retired_balance < Coin::from_u64_unchecked(Policy::MINIMUM_STAKE)
+        {
+            log::error!("{:?}", retired_balance);
+            return false;
+        }
+
+        true
+    }
+
+    /// Returns true if the inactive funds are released both from the cooldown period and any potential jail
+    /// of the current validator delegation.
+    fn is_inactive_stake_released(
+        &self,
+        store: &StakingContractStoreWrite,
+        staker: &Staker,
+        block_number: u32,
+    ) -> bool {
+        // If stake is currently delegated we check if funds are released.
+        // We only need to check cooldown of inactive funds if there is a delegation.
+        // This is because the purpose of the cooldown is to ensure that we wait for the reporting window
+        // to finish and thus that the validator will not be jailed for misbehavior.
+        if let Some(validator_address) = &staker.delegation {
+            // Funds are not released if release block height has not passed yet.
+            if let Some(inactive_from) = staker.inactive_from {
```

Comment:
However, if we have inactive funds delegated to nobody, we don't need to bother waiting for lockup or jail periods. 
The jail is self-explanatory, there is no delegation so the delegation cannot be jailed.
For the lockup checking it is unnecessary because the purpose of the lockup is to prevent the staker from retiring stake and potentially removing the stake before the reporting window for the misbehavior of a validator has passed. However, there is no delegation, so nobody can misbehave. Thus waiting for lockup is not needed.

Reply:
Ok :+1: